text/0000-transaction-refactoring.md

```diff
@@ -0,0 +1,508 @@
+
+# Motivation
+[motivation]: #motivation
+
+It simplifies account creation, by supporting batching of multiple transactions together instead of
+creating more complicated transaction types.
+
+For example, we want to create a new account with some account balance and one or many access keys, deploy a contract code on it and run an initialization method to restrict access keys permissions for a `proxy` function.
+
+To be able to do this now, we need to have a `CreateAccount` transaction with all the parameters of a new account.
+Then we need to handle it in one operation in a runtime code, which might have duplicated code for executing some WASM code with the rollback conditions.
+
+Alternative to this is to execute multiple simple transactions in a batch within the same block.
+It has to be done in a row without any commits to the state until the entire batch is completed.
+We propose to support this type of transaction batching to simplify the runtime.
+
+Currently callbacks are handled differently from async calls, this NEP simplifies data dependencies and callbacks by unifying them.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+### New transaction and receipts
+
+Previously, in the runtime to produce a block we first executed new signed transactions and then executed received receipts. It resulted in duplicated code that might be shared across similar actions, e.g. function calls for async calls, callbacks and self-calls.
+It also increased the complexity of the runtime implementation.
+
+This NEP proposes changing it by first converting all signed transactions into receipts and then either execute them immediately before received receipts, or put them into the list of the new receipts to be routed.
+To achieve this, NEP introduces a new message `Action` that represents one of atomic actions, e.g. a function call.
+`TransactionBody` is now called just `Transaction`. It contains the list of actions that needs to be performed in a single batch and the information shared across these actions.
+
+`Transaction` contains the following fields
+- `originator_id` is an account ID of the transaction originator.
+- `public_key` is to identify the access key used to signs the transaction.
+- `nonce` is used to deduplicate and order transactions.
+- `receiver_id` is the account ID of the destination of this transaction. It's where the generated receipt will be routed for execution.
+- `action` is the list of actions to perform.
+
+An `Action` can be of the following:
+- `CreateAccount` creates a new account. It has to be the first action, and the receiver has to be a new account. Until we have the account deletion, it doesn't make sense for trying to create an account after some other action. The action will fail if the account already exists.
+`CreateAccount` also grants permission for all subsequent batched action for the newly created account. For example, deploying code on the new account.
```

Comment:
We should clarify that for action `A_i` in list of actions `A_0,...,A_i,..., A_n` it has permission to operate only on the following account.

* As we execute actions we maintain the stack of account ids.
* Before we start executing them we pre-populate the stack with one account id -- the one that signed the transaction.
* Every `CreateAccount` puts new account id on the top of the stack;
* Every `DeleteAccount` removes account id from the stack;
* An action has rights to operate on the account id on the top of the stack.